src/main/java/com/aws/greengrass/FleetProvisioningByClaimPlugin.java

```diff
@@ -284,54 +293,104 @@ private void checkRequiredParameterPresent(Map<String, Object> parameterMap, Lis
         } catch (IOException e) {
-            logger.atError().log("Caught exception while writing certificate and private key to file");
-            throw new DeviceProvisioningRuntimeException("Failed to write certificate and private key", e);
+            logger.atError().kv("src", src).kv("dst", dst)
+                    .log("Caught exception while copying file");
+            throw new DeviceProvisioningRuntimeException(String.format("Failed to copy %s to %s", src, dst), e);
         }
     }
 
-    private void writeCertificateToPath(CreateCertificateFromCsrResponse response, String rootPath) {
+    private static void writeKey(CreateKeysAndCertificateResponse response, Path path) {
         try {
-            Path certPath = Paths.get(rootPath, DEVICE_CERTIFICATE_PATH_RELATIVE_TO_ROOT);
-            Files.createFile(certPath);
-            Files.write(certPath, response.certificatePem.getBytes(StandardCharsets.UTF_8));
-            Platform.getInstance().setPermissions(FileSystemPermission.builder().ownerRead(true)
-                .ownerWrite(true).build(), certPath);
+            writeFile(path, response.privateKey);
+            setFilePermissions(path);
         } catch (IOException e) {
             logger.atError().log("Caught exception while writing certificate to file", e);
             throw new DeviceProvisioningRuntimeException("Failed to write certificate", e);
         }
     }
+
+    private static void writeCertificate(CreateCertificateFromCsrResponse response, Path path) {
+        writeCertificate(response.certificatePem, path);
+    }
+
+    private static void writeCertificate(CreateKeysAndCertificateResponse response, Path path) {
+        writeCertificate(response.certificatePem, path);
+    }
+
+    private static void writeCertificate(String certificatePem, Path path) {
+        try {
+            writeFile(path, certificatePem);
+            setFilePermissions(path);
+        } catch (IOException e) {
+            logger.atError().log("Caught exception while writing certificate to file", e);
+            throw new DeviceProvisioningRuntimeException("Failed to write certificate", e);
+        }
+    }
+
+    private static void writeFile(Path file, String content) throws IOException {
+        createFile(file);
+        Files.write(file, content.getBytes(StandardCharsets.UTF_8));
+    }
+
+    private static void createFile(Path file) throws IOException {
+        if (file.getParent() != null) {
+            Files.createDirectories(file.getParent());
+        }
+        if (Files.notExists(file)) {
+            Files.createFile(file);
+        }
+    }
+
+    private static void setFilePermissions(Path file) {
+        try {
+            Platform.getInstance().setPermissions(FileSystemPermission.builder()
+                    .ownerRead(true)
+                    .ownerWrite(true)
+                    .build(), file);
+        } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Consider passing the exception object either as  &#34;rootCause&#34; or as an inner exception parameter, to the constructor of the new exception before throwing new exception. This will help keep stack trace.
```
@@ -392,3 +392,3 @@
         } catch (IOException e) {
-            throw new DeviceProvisioningRuntimeException(String.format(&#34;Failed to set permissions on file %s&#34;, file));
+            throw new DeviceProvisioningRuntimeException(String.format(&#34;Failed to set permissions on file %s&#34;, file), e);
         }

```